Add new list view - 'email'.

tim96 opened the pull request:
<!-- THE PR TEMPLATE IS NOT AN OPTION. DO NOT DELETE IT! -->

### Changelog

```markdown
### Added
- Added new field type `email` on the list
```

### Subject

Added new field type `email` on the list

### To do

- [X] Create PR
- [X] Add subject/body options 
- [X] Update the tests
- [x] Add info to changelog

Soullivaneuh commented on Resources/views/CRUD/list_email.html.twig:
```diff
@@ -0,0 +1,22 @@
+    {%- else -%}
+        <a href="mailto:{{ value }}">{{ value }}</a>
+    {%- endif -%}
+{% endblock %}
```
Please add a new line.

greg0ire replied:
In most cases, you should be able to configure your editor to automatically do so. It will make the red pictogram disappear.

OskarStark commented:
The `CHANGELOG.md` needs to be updated, too please (or am I wrong @Soullivaneuh )

OskarStark commented on DependencyInjection/AbstractSonataAdminExtension.php:
```diff
@@ -43,7 +43,7 @@ protected function fixTemplatesConfiguration(array $configs, ContainerBuilder $c
                         'text' => 'SonataAdminBundle:CRUD:list_string.html.twig',
                         'textarea' => 'SonataAdminBundle:CRUD:list_string.html.twig',
-                        'email' => 'SonataAdminBundle:CRUD:list_string.html.twig',
+                        'email' => 'SonataAdminBundle:CRUD:list_email.html.twig',
```
does this keep the same behaviour or is this a BC break?

Soullivaneuh replied:
What do you mean by "the same behaviour" ?

tim96 replied:
> does this keep the same behaviour or is this a BC break?

@OskarStark Instead of `admin@example.com` will be `<a href="mailto:admin@example.com">admin@example.com</a>`

OskarStark replied:
@Soullivaneuh this is not hard, but it is a BC break....

IMO we can merge it anyway in `3.x`

What do you say @greg0ire @core23 ?

greg0ire replied:
I say it would be a BC-break if the output was relied on by some program, but this is a listing, so only humans will be affected, and they will probably be thankful to have a link instead. :+1: to merge this in 3.x

OskarStark replied:
> I say it would be a BC-break if the output was relied on by some program, but this is a listing, so only humans will be affected, and they will probably be thankful to have a link instead. 👍 to merge this in 3.x

same here 👍

tim96 commented:
@OskarStark please wait some time, I will not finish yet. I don't know how to add label `in progress`.

OskarStark commented:
`in progress` is already on this PR

greg0ire commented:
Nice PR @tim96 !

Soullivaneuh commented:
> The CHANGELOG.md needs to be updated, too please (or am I wrong @Soullivaneuh )

No, we don't do that anymore because of conflicts issues: https://github.com/sonata-project/dev-kit/issues/70

tim96 commented:
How to remove 'in progress' tag?

greg0ire commented:
I don't think you can

greg0ire commented:
LGTM

Soullivaneuh commented on Resources/views/CRUD/list_email.html.twig:
```diff
@@ -0,0 +1,41 @@
+
+{% extends admin.getTemplate('base_list_field') %}
+
+{% block field%}
```
`{% block field%}` -> `{% block field %}` (missing a space before `%}`

Soullivaneuh commented on Resources/doc/reference/action_list.rst:
```diff
@@ -148,6 +148,12 @@ Available types and associated options
 +-----------+----------------+-----------------------------------------------------------------------+
 | datetime  | format         | A format understandable by Twig's ``date`` function.                  |
 +-----------+----------------+-----------------------------------------------------------------------+
+| email     | as_string      | Renders as email link                                                 |
```
The description is wrong too me.

If the `as_string` parameter is set to `true`, it deactivate the email link.

I would better write: 

```rst
Disables the email link if set to ``true``.
```

Soullivaneuh commented on Resources/doc/reference/action_list.rst:
```diff
@@ -148,6 +148,12 @@ Available types and associated options
 +-----------+----------------+-----------------------------------------------------------------------+
+| email     | as_string      | Renders as email link                                                 |
++           +----------------+-----------------------------------------------------------------------+
+|           | subject        | Add subject parameter to email link                                   |
```
Missing `.`

Soullivaneuh commented on Resources/doc/reference/action_list.rst:
```diff
@@ -148,6 +148,12 @@ Available types and associated options
++           +----------------+-----------------------------------------------------------------------+
+|           | subject        | Add subject parameter to email link                                   |
++           +----------------+-----------------------------------------------------------------------+
+|           | body           | Add body parameter to email link                                      |
```
Missing `.`

tim96 replied:
> Missing .

@Soullivaneuh Fixed.

Soullivaneuh commented:
@tim96 The PR is very good and could be accepted except my comments, thank you for your work! :+1: 

If I set this configuration:

```php
->add('email', 'email', [
    'as_string' => true,
    'subject' => 'TEST',
    'body' => 'This is a test.',
])
```

`subject` and `body` are useless because the link is not rendered. The best would be to throw a exception about it. But AFAIK, you can't because no option check.

It would be great to have an option resolver on next major, like it's done for form type.

Soullivaneuh commented on Resources/views/CRUD/list_email.html.twig:
```diff
@@ -0,0 +1,41 @@
+        {%- set url = '' -%}
+        {%- if field_description.options.subject is defined and field_description.options.subject is not empty -%}
+            {%- set subject = field_description.options.subject -%}
+            {%- set url = 'subject=' ~ subject -%}
```
Code between `list_email.html.twig` and `show_email.html.twig` is mostly the same.

It would be great to avoid duplication with code refactoring.

Two solutions:

1. A common template like `_value_email.html.twig` (not sure for the name) which can be included
2. A twig extension. Something like `value | mailto_link(subject, body)`.

For options `2`, we have to decide where to put the extension. Admin? Core? Maybe an external library?

BTW, maybe a package exists providing useful twig extension like this.

tim96 replied:
> Code between list_email.html.twig and show_email.html.twig is mostly the same.

I read almost all templates for list variables like `list_url`, `list_date` and etc. Almost all templates have duplication code, so I try to keep save code style.

> Two solutions:

I have multiple ideas, I will check it and write about it.

Soullivaneuh replied:
> I read almost all templates for list variables like `list_url`, `list_date` and etc. Almost all templates have duplication code, so I try to keep save code style.

I understand. But this is not because we did some things in the past than we should not improve them. :wink:

tim96 replied:
> A common template like _value_email.html.twig (not sure for the name) which can be included

This solution is working I think. @Soullivaneuh I can add it for this PR

Soullivaneuh replied:
Feel free to push your proposition! :+1: 

I suggest you to make it on a separate commit. It would be easier for you to remove it if we decide for another solution.

Soullivaneuh commented on Resources/doc/reference/action_list.rst:
```diff
@@ -148,6 +148,12 @@ Available types and associated options
 +-----------+----------------+-----------------------------------------------------------------------+
 | datetime  | format         | A format understandable by Twig's ``date`` function.                  |
 +-----------+----------------+-----------------------------------------------------------------------+
+| email     | as_string      | Renders as email link.                                                |
```
Please change the description: `Renders the email as string, without any link.`

Soullivaneuh commented on Resources/views/CRUD/_email_link_template.html.twig:
```diff
@@ -0,0 +1,29 @@
+
+{% block email_link_template %}
```
I don't think the `block` is useful here.

Soullivaneuh commented:
`_email_link_template.html.twig` -> `_email_link.html.twig`?

cc @sonata-project/contributors

core23 commented:
Maybe `base_email_link.html.twig`?

OskarStark commented:
> Maybe base_email_link.html.twig?

👍

Soullivaneuh commented:
It's a partial template. Partial template should start with `_` IMO.

greg0ire commented:
> Partial template should start with _ IMO.

Is this a best practice from symfony1 ?

Soullivaneuh commented:
@greg0ire I don't know. But this is easier to identify partial template, isn't it?

greg0ire commented:
> @greg0ire I don't know. But this is easier to identify partial template, isn't it?

Well yes if you know the meaning, but I did not see this naming convention since sf1. Can you find any sf2 docs that recommend doing that?

Soullivaneuh commented:
As I said, I don't think so. This is just an opinion.

greg0ire commented:
Ok so :-1: on this :P

javiereguiluz commented:
@greg0ire this convention is not written in the Symfony docs but it's well known and well followed by lots of Symfony 2 and 3 developers. In fact, the official Symfony Demo app follows it too. See for example https://github.com/symfony/symfony-demo/tree/master/app/Resources/views/blog

Soullivaneuh commented:
> this convention is not written in the Symfony docs but it's well known and well followed by lots of Symfony 2 and 3 developers.

@greg0ire so :+1: on this. :stuck_out_tongue: 

@javiereguiluz It would be great is this can be included on the SF best practices. :+1:

greg0ire commented:
@javiereguiluz : O_o ok then… looks a lot like hungarian notation, but if the Symfony documentation does it… never noticed it before though.

greg0ire commented on Resources/views/CRUD/_email_link.html.twig:
```diff
@@ -0,0 +1,27 @@
+    {%- if field_description.options.subject is defined and field_description.options.subject is not empty -%}
+        {%- set subject = field_description.options.subject -%}
+        {%- set url = 'subject=' ~ subject -%}
+    {%- endif -%}
```
It think you could simplify all this greatly by using [default](http://twig.sensiolabs.org/doc/filters/default.html):

Maybe try something like `{% set url = field_description.options.subject|default('') %}`

Soullivaneuh replied:
:+1: 

@tim96 can you please update your PR?

tim96 replied:
@greg0ire Thank you for idea.

> @tim96 can you please update your PR?

@Soullivaneuh Fixed

Soullivaneuh commented on Resources/views/CRUD/_email_link.html.twig:
```diff
@@ -0,0 +1,21 @@
+
+{%- if value is empty -%}
+    &nbsp;
```
Is this really necessary?

cc @greg0ire

tim96 replied:
> Is this really necessary?

@Soullivaneuh I saw something like this in multiple files, for example:
`list_datetime.html.twig`
`list_date.html.twig`
`list_html.html.twig`

Soullivaneuh replied:
Well, OK then. But I think this should be reviewed later.

Soullivaneuh commented:
LGTM, thanks @tim96! :+1:

OskarStark commented:
LGTM, too 👍

Soullivaneuh commented:
@tim96 Travis is failind for PHP 5.3, could you please take a look?

tim96 commented:
> @tim96 Travis is failing for PHP 5.3, could you please take a look?

@Soullivaneuh at first look it seems that function `url_encode` create diff results for version php 5.3 and php 5.4

OskarStark commented:
> @Soullivaneuh at first look it seems that function url_encode create diff results for version php 5.3 and php 5.4

if this is the problem, i think we can only merge this in the next major

greg0ire commented:
> @Soullivaneuh at first look it seems that function url_encode create diff results for version php 5.3 and php 5.4

Well sorry for that, moving it to the next major might be best indeed, because the code is simpler…

Soullivaneuh commented:
Or maybe use [`rawurlencode`](http://php.net/manual/fr/function.rawurlencode.php)?

BTW @tim96 the code coverage decreased, could you please take a look?

tim96 commented:
I found this information:
http://php.net/manual/en/function.http-build-query.php
```
 enc_type
    By default, PHP_QUERY_RFC1738.
    If enc_type is PHP_QUERY_RFC1738, then encoding is performed per » RFC 1738 and the application/x-www-form-urlencoded media type, which implies that spaces are encoded as plus (+) signs.
    If enc_type is PHP_QUERY_RFC3986, then encoding is performed according to » RFC 3986, and spaces will be percent encoded (%20).
```
So I think it is right behaviour, so we need simple update test case for this situation. What do you think @Soullivaneuh ?

> BTW @tim96 the code coverage decreased, could you please take a look?

Some thing wrong with this program, because I don't add new methods. Maybe I need to add test case for this method but this method using only to provide some data for another test case. It is very strange and I didn't work with this website before, sorry((

greg0ire commented:
@tim96 : we introduced coveralls recently, and it might not have coverage for the commit you are starting from, which means the solution for it to behave normally would be, you probably guessed it… a rebase!

Soullivaneuh commented:
> Some thing wrong with this program, because I don't add new methods.

Indeed. As @greg0ire said, try a rebase. :+1: 

> So I think it is right behaviour, so we need simple update test case for this situation. What do you think @Soullivaneuh ?

Do the failing tests use `http_build_query`? In this case, yes I think. But use the default of the most recent php version.

tim96 commented:
> Do the failing tests use http_build_query? In this case, yes I think. But use the default of the most recent php version.

Yes and no. In similar test cases sonata don't use test examples with 'space' symbols, so tests would not fail on php 5.3 version. So I can remove such sentence and pass all tests, for example instead of `Main Theme` using `Main`. @Soullivaneuh I add one function in tests, but I am not sure about right way.

> a rebase!

@greg0ire Fixed.

greg0ire commented on Tests/Twig/Extension/SonataAdminExtensionTest.php:
```diff
@@ -340,6 +342,16 @@ public function testDeprecatedRenderListElement($expected, $value, array $option
+    public function buildUrl($url)
+    {
+        if (defined('PHP_QUERY_RFC3986')) {
+            // add htmlspecialchars because twig add it auto
```
I don't understand…

tim96 replied:
Since php5.4 exist const 'PHP_QUERY_RFC3986' and twig using it in url_encode.

```
/**
 * URL encodes (RFC 3986) a string as a path segment or an array as a query string.
 *
 * @param string|array $url A URL or an array of query parameters
 *
 * @return string The URL encoded value
 */
function twig_urlencode_filter($url)
{
    if (is_array($url)) {
        if (defined('PHP_QUERY_RFC3986')) {
            return http_build_query($url, '', '&', PHP_QUERY_RFC3986);
        }

        return http_build_query($url, '', '&');
    }

    return rawurlencode($url);
}
```

What's why my test case fail on php 5.3

@greg0ire please look https://3v4l.org/sX7Dn

greg0ire replied:
What I don't understand is the comment : why add htmlspecialchars if twig adds it automatically?

tim96 replied:
If you want show some html code you need using `raw` option otherwise it will replace:
```
{{ {foo: "bar", number: 3}|url_encode }}
{{ {foo: "bar", number: 3}|url_encode|raw }}
The results:
foo=bar&amp;number=3
foo=bar&number=3

```

greg0ire replied:
Oh ok I get it now! For some reason I thought this was in a new Twig filter you wrote… I must be tired!

greg0ire replied:
I guess both outputs are ok so this must be the right way indeed! Please mark your method with

`// NEXT_MAJOR: remove this method`. We will remove php 5.3 support in next major, so we won't need it anymore.

Soullivaneuh commented on Tests/Twig/Extension/SonataAdminExtensionTest.php:
```diff
@@ -340,6 +342,16 @@ public function testDeprecatedRenderListElement($expected, $value, array $option
         );
     }
 
+    public function buildUrl($url)
```
* Call it `buildTwigLikeUrl`
* Make it `private`
* A a phpdoc
* On the PHP doc, specify that is a BC for PHP 5.3 and twig
* As @greg0ire said, add something like `NEXT_MAJOR: remove this method as PHP 5.3 support will be dropped`

tim96 replied:
> Call it buildTwigLikeUrl

@Soullivaneuh Thank you for review. Fixed.

Soullivaneuh commented:
Nearly OK @tim96, thanks for your effort! :+1:

tim96 commented:
> @tim96 : we introduced coveralls recently, and it might not have coverage for the commit you are starting from, which means the solution for it to behave normally would be, you probably guessed it… a rebase!

@greg0ire hi. I rebased my PR but I'm not sure about test coverage.

Soullivaneuh commented on Tests/Twig/Extension/SonataAdminExtensionTest.php:
```diff
@@ -2002,6 +2130,25 @@ public function testGetUrlsafeIdentifier_GivenAdmin_Bar()
         $this->assertSame(1234567, $this->twigExtension->getUrlsafeIdentifier($entity, $this->adminBar));
     }
 
+    // TODO: remove this method for next major release only if PHP 5.3 support will be dropped
```
No `TODO` allowed on the code.

Just move you sentence below the first on on the phpdoc of `buildTwigLikeUrl`. :+1:

tim96 replied:
> Just move you sentence below the first on on the phpdoc of buildTwigLikeUrl. :+1:

Done

Soullivaneuh commented on Tests/Twig/Extension/SonataAdminExtensionTest.php:
```diff
@@ -2002,6 +2130,24 @@ public function testGetUrlsafeIdentifier_GivenAdmin_Bar()
     }
 
+    /**
+     * Remove this method for next major release only if PHP 5.3 support will be dropped
```
After the line, not before. :stuck_out_tongue: 

Also add a blank line between the two to have a clearer phpdoc.

And add a dot at the end, or StyleCI will cry. :wink:

tim96 replied:
> And add a dot at the end, or StyleCI will cry. :wink:

@Soullivaneuh very strange but StyleCI will not cry

Soullivaneuh commented on Tests/Twig/Extension/SonataAdminExtensionTest.php:
```diff
@@ -2002,6 +2130,24 @@ public function testGetUrlsafeIdentifier_GivenAdmin_Bar()
 
+    /**
+     * Remove this method for next major release only if PHP 5.3 support will be dropped
+     * This method generate url part for twig layout. Allow to keep BC for PHP 5.3.
```
* `generate` -> `generates`
* `for twig` -> `for Twig`
* `Allow` -> `Allows`

tim96 replied:
> 

    generate -> generates
    for twig -> for Twig
    Allow -> Allows

@Soullivaneuh Fixed

tim96 commented on Tests/Twig/Extension/SonataAdminExtensionTest.php:
```diff
@@ -2002,6 +2130,25 @@ public function testGetUrlsafeIdentifier_GivenAdmin_Bar()
+     *
+     * Remove this method for next major release only if PHP 5.3 support will be dropped.
+     *
+     * @param $url array
```
> @param array $url

Done.

Soullivaneuh commented:
RTM! :tada: 

@tim96 Thanks a lot, again. :+1:
